contracts/relayer/Relayer.sol

```diff
@@ -0,0 +1,349 @@
+
+    /**
+     * @notice Tell if a given sender `sender` is allowed to use the relay service.
+     * @return True if the given sender is allowed to use the relay service
+     */
+    function isSenderAllowed(address sender) external view isInitialized returns (bool) {
+        return _isSenderAllowed(sender);
+    }
+
+    /**
+     * @notice Tell if a given sender `sender` can use the nonce number `nonce` to relay a new transaction.
+     * @return True if the given sender can use the given nonce number to relay a new transaction
+     */
+    function canUseNonce(address sender, uint256 nonce) external view isInitialized returns (bool) {
+        return _canUseNonce(sender, nonce);
+    }
+
+    /**
+     * @notice Tell if a given sender `sender` can relay a new transaction spending `@tokenAmount(newQuota, 0x00)` in month `month`.
+     * @return True if the given sender can relay a new transaction spending the given amount for the given month
+     */
+    function canRefund(address sender, uint256 month, uint256 amount) external view isInitialized returns (bool) {
+        return _canRefund(sender, month, amount);
+    }
+
+    /**
+     * @notice Tell if the current app allows to recover amount of `token` from the Relayer app.
+     * @param token Token address that would be recovered
+     * @return True if the given address is not ETH
+     */
+    function allowRecoverability(address token) public view returns (bool) {
+        // does not allow to recover ETH
+        return token != ETH;
+    }
+
+    function _getCurrentMonth() internal view returns (uint256) {
+        uint256 passedSeconds = getTimestamp().sub(startDate);
+        return passedSeconds / 30 days;
+    }
+
+    function _getLastUsedNonce(address sender) internal view returns (uint256) {
+        return lastUsedNonce[sender];
+    }
+
+    function _getMonthlyRefunds(address sender, uint256 month) internal view returns (uint256) {
+        return monthlyRefunds[sender][month];
+    }
+
+    function _isServiceAllowed(address service) internal view returns (bool) {
+        return allowedServices[service];
+    }
+
+    function _isSenderAllowed(address sender) internal view returns (bool) {
+        return allowedSenders[sender];
+    }
+
+    function _canUseNonce(address sender, uint256 nonce) internal view returns (bool) {
+        return _getLastUsedNonce(sender) < nonce;
+    }
+
+    function _canRefund(address sender, uint256 month, uint256 amount) internal view returns (bool) {
+        uint256 monthRefunds = _getMonthlyRefunds(sender, month);
+        return monthRefunds.add(amount) <= monthlyRefundQuota;
+    }
+
+    function _isValidSignature(address sender, address to, uint256 nonce, bytes data, uint256 gasRefund, uint256 gasPrice, bytes signature)
+        internal
+        view
+        returns (bool)
+    {
+        bytes32 messageHash = _messageHash(to, nonce, data, gasRefund, gasPrice);
+        address signer = messageHash.recover(signature);
```

Comment:
No, that's the idea of supporting [EIP 712](https://eips.ethereum.org/EIPS/eip-712) 😉

Reply:
I thought it may be related to it, but didn't read all the details of that EIP, and I was confused by the fact that `toEthSignedMessageHash` is still in ECDSA lib. Do we need it there then?

Reply:
Not really, it is part of the library I picked from OpenZeppelin